### ISSUE TITLE
ExecutePreprocessor.timeout do not work properly

### ISSUE THREAD

**Issue by MrBanhBao** [jupyter/nbconvert#857]
I currently trying to run a notebook via the ExecutePreprocessor (Python API) after I've trained a Keras image-classification model.
The notebook just calculates some metrics and plots wrong predicted images with matplotlib.

On my dummyset which consists of 40 imgs the preprocessor is doing fine.
After running the notebook on a bigger image set a TimeoutEroor was raised.

As result I set the timout parameter of the ExecutePreprocesor to None / -1 and also 1800 but the problem still consist.

Code:
`def run_jupyter_notebook(model_path, base_model_name, test_img_dir, save_path):
    with open('evaluate_notebook.ipynb') as f:
        nb = nbf.read(f, as_version=4)

    # Get a list of Parameter objects
    orig_parameters = extract_parameters(nb)

    # Update one or more parameters
    params = parameter_values(orig_parameters,
                              model_path=model_path,
                              test_img_dir=test_img_dir)

    new_nb = replace_definitions(nb, params)

    ep = ExecutePreprocessor(timeout=None)

    ep.preprocess(new_nb, {'metadata': {'path': '.'}})
    nbf.write(new_nb, save_path)`

Error Messsages:
`2018-08-10 15:02:31.883751: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
Traceback (most recent call last):
  File "/home/hao/anaconda3/envs/img-class/lib/python3.5/site-packages/nbconvert/preprocessors/execute.py", line 324, in _wait_for_reply
    msg = self.kc.shell_channel.get_msg(timeout=timeout)
  File "/home/hao/anaconda3/envs//img-class/lib/python3.5/site-packages/jupyter_client/blocking/channels.py", line 57, in get_msg
    raise Empty
queue.Empty

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/hao/anaconda3/envs/hotel-image-assessment/lib/python3.5/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/home/hao/anaconda3/envs/img-class/lib/python3.5/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/hao/idealo/hotel-image-tagging/src/trainer/train.py", line 135, in <module>
    main(**config, **args.__dict__)
  File "/home/hao/idealo/hotel-image-tagging/src/trainer/train.py", line 86, in main
    '../../outputs/' + base_model_name + '/' + ts + '/' + 'eval_goldset.ipynb')
  File "/home/hao/idealo/hotel-image-tagging/src/evaluate/eval.py", line 19, in run_jupyter_notebook
    new_nb = replace_definitions(nb, params)
  File "/home/hao/anaconda3/envs/img-class/lib/python3.5/site-packages/nbparameterise/code.py", line 98, in replace_definitions
    nb, resources = ExecutePreprocessor().preprocess(nb, resources)
  File "/home/hao/anaconda3/envs/img-class/lib/python3.5/site-packages/nbconvert/preprocessors/execute.py", line 262, in preprocess
    nb, resources = super(ExecutePreprocessor, self).preprocess(nb, resources)
  File "/home/hao/anaconda3/envs/img-class/lib/python3.5/site-packages/nbconvert/preprocessors/base.py", line 69, in preprocess
    nb.cells[index], resources = self.preprocess_cell(cell, resources, index)
  File "/home/hao/anaconda3/envs/img-class/lib/python3.5/site-packages/nbconvert/preprocessors/execute.py", line 280, in preprocess_cell
    reply, outputs = self.run_cell(cell, cell_index)
  File "/home/hao/anaconda3/envs/img-class/lib/python3.5/site-packages/nbconvert/preprocessors/execute.py", line 348, in run_cell
    exec_reply = self._wait_for_reply(msg_id, cell)
  File "/home/hao/anaconda3/envs/img-class/lib/python3.5/site-packages/nbconvert/preprocessors/execute.py", line 337, in _wait_for_reply
    raise exception("Cell execution timed out")
TimeoutError: Cell execution timed out
`

**Comment by akhmerov** [comment]
What version of nbconvert do you have?
Do you also observe the problem on a minimal notebook containing
```python
from time import sleep; sleep(3600)
```
?
Can you recheck that `timeout=-1` does not work (`timeout=None` means "use default")?

**Comment by akhmerov** [comment]
I double-checked on the notebook that sleeps for an hour, and I cannot reproduce:
```
jupyter nbconvert --execute --ExecutePreprocessor.timeout=-1 --to notebook Untitled.ipynb                                                                         (main) 
[NbConvertApp] Converting notebook Untitled.ipynb to notebook
[NbConvertApp] Executing notebook with kernel: python3
[NbConvertApp] Writing 606 bytes to Untitled.nbconvert.ipynb
```

**Comment by MrBanhBao** [comment]
@akhmerov thanks for your fast reply
my nbconvert version is 5.3.1.

I ran the jupyter notebook via terminal with your command with two different bigger datasets (106 and 502 images) by providing the paths manually inside the notebook and it worked.

But I want to run the notebook with the Python API of nbconvert. Therefore I wrote a function with parameter values which will replace static variables inside the notebook via nbparameterized (see code in original post).
Calling the notebook via Python API just works with the small dummy dataset (40 images) for me, running the notebook with datasets with datasets with 106 and 502 images do not work, it throws a TimeoutError.

That's why I think that setting the timeout paramter via Python API doesn't work properly.
`ep = ExecutePreprocessor(timeout=-1)`

**Comment by MrBanhBao** [comment]
@akhmerov you were right I ran your code without nbparameterise methods and it worked fine.
I figured out that replace_definitions method from nbparameterise also runs the notebook with the default timeout, so the error message was produced by replace_definition.

After setting execute argument of to false it the notebook ran without throwing any errors.
`replace_definitions(nb, params, execute=False)`

Thank you for your help!
I think this issue can be closed.
